### ISSUE TITLE
Inheritance memory leak when upgrading and destructing in the same frame

### ISSUE THREAD

**Issue by nyankers** [dworkin/dgd#44]
This was pretty tough to nail down why it was happening, but it plagued my little development server with a few stray library issues after I wrote my own code command with extra bells and whistles. Cloud Server should be fine, as it takes necessary precautions to prevent this (destroying the object before creating ensures it doesn't go into upgrade state, mine was missing this step but still destructed afterwards, but both code commands can fail to reach this point if the output exceeds string max size)... but it turns out this issue can happen with user-level code too!

First, you'll want to create a new user so you have handy access to the `issues` command (`upgrade` is optional, you can recreate this with `destruct`/`compile`). I did this via `grant test access` and `grant test / full` and then logging in as `test`.

Then create two files in the new user's /usr/ directory:

test.c:
```
void test()
{
	if (!find_object("test2")) {
		compile_object("test2");
	}
	compile_object("test2");
	destruct_object("test2");
}
```

test2.c (this can be whatever, as long as it compiles), e.g.:
```
void f()
{
}
```

Run `compile test.c` to compile it, then run the following commands in sequence to create the zombie process, you can do so multiple times to see more and more issues get added:
```
code "test" -> test()
upgrade /usr/System/lib/auto.c
issues /usr/System/lib/auto.c
```

I don't know what the correct solution is, but I can at least explain what's going on. Basically, the second `compile_object()` will initiate an upgrade for the object, so `O_UPGRADING(x)` will return true for it. Then we're immediately deleting that object, calling `Object::del()` which removes the references for our newly created object right back down to 0, but since it's marked as upgrading, it refuses to call `remove(f)`. I'd guess there should be some follow-up task to actually call `remove(f)` (or maybe the upgrade check itself is wrong), but that never happens, so the library gets the added references from being compiled against, but they're never taken away.

As a follow-up question, is there any way to fix the state once it's broken by something like this? For example, some way of getting dgd to recount inheritance references? Since I only have a few (around 17), I suppose I could manually go in with GDB and fiddle with the ref counts directly.

**Comment by nyankers** [comment]
It looks like this might also be causing Object::clean() to skip freeing up some memory. I also tried to find a reason for that upgrade check in Object::del(), but I don't see one. Obviously you'd know better than I would, but if it's any help, I'll submit a pull request to remove it that demonstrates precisely what I mean.

After it's removed, Object::clean() seems to handle the object more correctly in this situation, and no zombie lib issues are created.

**Comment by dworkin** [comment]
This was specifically added in commit dea1f390a0a6f98aed25aaa4290fe667e73f71ce. I have to look more deeply into what that fixed, and how to make sure that doesn't happen anymore.

**Comment by dworkin** [comment]
Once again, your analysis and solution are correct.

This was a real puzzler. In 2013, I worked for a while on a branch of Hydra in which it was possible to recompile inherited objects. The branch was never integrated, but I cherry-picked fixes for bugs I uncovered to the master branch, erroneously including this bad change which only made sense on the originating branch. That change was then copied to DGD, as well.

**Comment by dworkin** [comment]
How to fix your snapshot. The best way might be to create a function which searches the object table for left-over test2 objects, and then calls `remove` for them.

**Comment by shentino** [comment]
Btw is there any chance this issue could have corrupted other snapshots in general?  I'm currently running a live version of kotaka on gcp as an exercise in server management

**Comment by nyankers** [comment]
Ah, for some reason I mistakenly thought that objects handled by `Object::del()` would free up the object table, but checking again that's also handled by `Object::remove()`.

@shentino It looks like this code went in version 1.4.20, so any persistence server using that version or later could be affected until the above patch, I'd assume based on the above response that it was erroneous from the start. That said, you'd only be afflicted if you ever ran code which upgraded and destructed an object in the same frame, which is a pretty unusual thing to do intentionally and only happened to me because I wrote a bad `code` command that sometimes did exactly that.

@dworkin Based on that, should there be something like an upgrade task to automatically clean up any such objects? I could try to cobble something up for it later, but the usefulness of that would entirely rely on you accepting it, and my personal needs don't need much more than a free-floating function that does the same (which I'd be happy to make available as well, but the level of support there would be "here's the code, good luck!")

**Comment by nyankers** [comment]
By the way, my assumption currently is that these zombie objects (e.g. `test2` in the above example) can be identified by having `count == 0 && ref == 0 && name != NULL`, provided that they're not already on the ObjPlane's destruct list (sort of an important and slightly annoying condition as running `Object::remove()` will potentially put inherited objects on there as well).

**Comment by nyankers** [comment]
On further inspection, the upgraded `Control` seems to be lost (meaning the object's new inherits list is also lost), so a general patch would have to do more than just intelligently call `Object::remove()`. Instead, it may have to look more like this:
1. Collect all the zombie objects (e.g. `test2`) that need to be cleaned up.
2. Call driver->`remove_program` on them all, and add them to the destruct pile.
3. Recalculate every non-zombie object's reference count.
4. Call `Object::remove()` on every non-zombie object that step 3 found should be removed.

Assuming I'm not missing anything else...

**Comment by dworkin** [comment]
Zombie objects might indeed not even have a control block.  What I've done in the past in cases of snapshot corruption is write a quick temporary patch rather than something like a full consistency check.

Since you are working with cloudlib, at least you can easily identify the spurious `~System/lib/auto` issues easily, and then call `remove` for them. The leftover zombie objects would show up as having reference count 0 without being in the free object list.

**Comment by nyankers** [comment]
They apparently do not have a control block. I created such a patch that works roughly as I described above over here: https://github.com/nyankers/dgd/tree/remove-zombies which github has already sort of pointed out.

Even in my environment, it didn't quite catch everything though (it found one object with no name, no count, cref way bigger than ref, but ref was 1 and none of the baseplane's lists pointed to it including free, and also objectd doesn't know it "exists"), so I can't guarantee it won't break if anyone else wants to try it, as my environment is pretty small and simple.
